Transpile examples

ttodua opened the pull request:
PR aims to automatically transpile `/example/ts` files.

- to only transpile the target files, we have to include "AUTO-TRANSPILE" (or any phrase you want) in files, which we want to be transpiled.  (in future, we can change that approach).
- catch & recognize all exchange instantiations within functions and close them in async python with `await exchange.close`. 
- you can now name function whatever you want.
- you can now have many functions (calling one from another)
- you can instantiate as many exchanges in function, as you want, and all of their variable names will be detected and will be closed with `await instanceName.close()`.

iIt was tested against `fetch-ohlcv.js` example file. waiting your notes.

carlosmiei commented on examples/js/fetch-ohlcv.js:
```diff
@@ -0,0 +1,22 @@
+// AUTO-TRANSPILE //
+
+async function myfunc () {
+    const exchange = initExchange ('binance');
```
not sure if this is flexible enough to handle things like custom options, credentials etc 🤔

ttodua replied:
absolutely that's not flexible enough. that's why i said in top post "this is just start" and any further notes & edits are needed. i'll update it now but plz add your suggestions too. again, we should start from something to make it grow.

carlosmiei replied:
@ttodua sure this is already a great start 🙂 

second question, should we consider transpile this to python/php async?

ttodua replied:
fixed the transpiler, so there is no need to be separate `initExchange` (the class instantiation is now being transpiled natively)

ttodua replied:
about async - i need your thoughts, how we should do it. there are basic examples, that only calls a single endpoint (and in those cases, there is no need of async imho), however, for the sake of the completeness, should we transpile every js file, i.e.:
`filename.js`
--> `filename.php` & `filename-async.php`
--> `filename.py` & `filename-async.py`
?

ttodua replied:
also, there are some gotchas, i.e. in js you can call `async` method without `await`:
```
async function x() {

}
x();
```
however, in py it needs to have `await x()` otherwise throws error. So, we are not fully free to write JS examples in such manner. or do you think of workaround?

ttodua replied:
added also async-php, however, both py/php are structurally different and this will need more edits to become better. However, at this moment, as you can test it out too, it does work with `fetch-ohlcv.js` example.

carlosmiei replied:
> --> filename.php & filename-async.php
@ttodua, actually I think we should place all transpiled examples in a different folder, what do you think? By doing that it's easy to tell if some example is automatically transpiled or not

> example() to be name of the sample function
humm why is that? 

> exchange variable to be the main holder of instantiated
yeah this is because of the .close right? yeah for this  I don't see a workaround

ttodua replied:
@carlosmiei I don't know which is best - to mix the transpilable snippets in main JS folder with others, or not. 
imo, it can be in same folder, as is, and the `AUTO-TRANSPILE` head-comment makes it clear for users, that the file is transpiled one.
@kroitor your thoughts are needed.

ttodua replied:
2)

ttodua replied:
it also needs some brainstorming, to think how to design the rules for writing the transpilable example. to refer to `await ex.close()`  problem in py, or async method recognition and their correct transpilation into py/php (as their async syntax is different from JS async/await flow...

ttodua commented:
@kroitor @carlosmiei after fiddling with this whole day, finally I was able to ("so so") make transpiler more flexible to : 
- catch & recognize all exchange instantiations within functions and close them in async python with `await exchange.close`. 
- removed the need to be `example` as function name - you can now name function whatever you want.
- removed the need to be only one function in example - you can now have many functions (calling one from another)
- removed the need to be `exchange` as variable name - you can name initiated exchange holder variable whatever name you want, moreover, you can instantiate as many exchanges in function, as you want, and all of their variable names will be detected and will be closed with `await varName.close()`.

this is not still an ideal solution, but I think we can start from this. waiting you thoughts and tests. It was tested against `fetch-ohlcv.js` example file.

carlosmiei commented:
@ttodua thanks, you did a really good job here

Couple of questions that I have:
- Do we need to create a sync/async version in Python/PHP for every example? IMO the async one should be enough, and we save some memory and build time
- Do we need to transpile it in every build? I think this should be a separate command like `npm run transpile-examples` or something similar, 99% of the time we build is not to add a new example, it can be a bit overkill.

carlosmiei commented on examples/js/fetch-ohlcv.js:
```diff
@@ -0,0 +1,18 @@
+'use strict';
+const ccxt = require ('../../ccxt.js');
```
usually end users don't import ccxt like this

ttodua replied:
Thanks! 
A) I can't say decisively, whether we need async & sync examples. I just come out from the point, that some users either use only sync version, thus 'async-only' examples wouldn't be help for them imo. if we just make `sync` only examples, then ppl which only use async version, will need async-example. So, in my current view, both versions are needed.
B) ofc, it's not needed to re-transpile it on every build. I didn't have implemented that, but ofc likewise we do derived exchange classes, we can cache/rebuild it only whenever there is change. But I dont think that caching mechanism would be in this same PR's scope, because :
- for local builds - yes, we can just make it like we do for derived exchange files (only transpiling the changed ones. I'll add commit for that).
- for travis builds - it might be slightly trickier than being said , bcz we will need somehow to make travis aware of previous states, which will need some local cache or etc.. So, maybe for this current moment, we skip caching issue and we can address that later?

ttodua replied:
ah, I've just copied that line from another example-js.

ttodua replied:
btw, how you suggest this to be? so the example files were runnable directly.

ttodua commented:
Thanks! 
A) I can't say decisively, whether we need async & sync examples. I just come out from the point, that some users either use only sync version, thus 'async-only' examples wouldn't be help for them imo. if we just make `sync` only examples, then ppl which only use async version, will need async-example. So, in my current view, both versions are needed.
B) ofc, it's not needed to re-transpile it on every build. I didn't have implemented that, but ofc likewise we do derived exchange classes, we can cache/rebuild it only whenever there is change. But I dont think that caching mechanism would be in this same PR's scope, because :
- for local builds - yes, we can just make it like we do for derived exchange files (only transpiling the changed ones. I'll add commit for that).
- for travis builds - it might be slightly trickier than being said , bcz we will need somehow to make travis aware of previous states, which will need some local cache or etc.. So, maybe for this current moment, we skip caching issue and we can address that later?

ttodua commented:
@carlosmiei sorry if this made a trouble for you, but i've completely re-organized the syntax (not functionality itself, just variable names and format) of the code, because we should aim to be more type/key dependent because of many flat-level variables, instead of defining separate variable names for those flat-level variables.

ttodua commented:
@carlosmiei if you notice, the `force` flag (which handles files re-transpilation) is only present/passed to [`transpileDerivedExchangeClasses`](https://github.com/ccxt/ccxt/blob/11a104be88f4585d353811da5fe1c8cfe456285f/build/transpile.js#L2052 ), but not passed in other methods -  `this.transpileTests, transpileExamples, transpilePythonAsyncToSync, transpilePhpAsyncToSync, transpilePhpBaseClassMethods ..` , so maybe that caching/re-transpilation thing needs to be addressed in a separate PR for each of those missing methods.

ttodua commented:
@frosty00 I saw you were able to merge few PR's related to examples. thanks! would be good if you might also give one step further and revise this PR - also purely related to `examples` . This PR  is not intrusive and only adds 1 function inside transpiler.

ttodua commented:
Merged, good !
I'll move now some other files into ts folder.
